juju/juju
# audit-logging: integrate auditlog.Recorder into API server

## Description of change

This hooks up the core/auditlog package to the API server, by creating a recorder in the Admin facade when the user logs in and having the server `rpc.Conn` notify it about requests and replies.

Replaces the observer factory the Conn was using with a `RecorderFactory` - the only real difference is that the `Recorder.ServerRequest` and `.ServerReply` methods allow returning an error (so we don't allow an operation to proceed if we can't write it to the audit log). Then the API server passes in a `Recorder` that combines the observers with an audit logger.

At the moment the audit logging always captures method arguments, and doesn't extract errors from the responses. Those will come in subsequent PRs.

## QA steps
Spin up a controller with `--auditing-enabled=true`. Check that messages corresponding to user commands like `deploy` and `add-unit` are logged in `/var/log/juju/audit.log`, as well as the resulting API calls.

## Documentation changes
The audit logging will need documentation.
